pkg/api/vm/handler.go

```diff
@@ -1367,3 +1390,53 @@ func convertNodeSelectorRequirementToSelector(req corev1.NodeSelectorRequirement
 		return &labels.Requirement{}, fmt.Errorf("unsupported operator: %v", req.Operator)
 	}
 }
+
+func (h *vmActionHandler) updateResourceQuota(namespace, name string, input UpdateResourceQuotaInput) error {
+	totalSnapshotSizeQuotaQuantity, err := resource.ParseQuantity(input.TotalSnapshotSizeQuota)
+	if err != nil {
+		return apierror.NewAPIError(validation.InvalidBodyContent, fmt.Sprintf("Failed to parse totalSnapshotSizeQuota: %v", err))
+	}
+	totalSnapshotSizeQuotaInt64, ok := totalSnapshotSizeQuotaQuantity.AsInt64()
+	if !ok {
+		return apierror.NewAPIError(validation.InvalidBodyContent, "Failed to parse totalSnapshotSizeQuota as int64")
+	}
+
+	resourceQuota, err := h.resourceQuotaClient.Get(namespace, util.DefaultResourceQuotaName, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if totalSnapshotSizeQuotaInt64 == 0 {
+				return nil
+			}
+			resourceQuota = &harvesterv1.ResourceQuota{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: namespace,
+					Name:      util.DefaultResourceQuotaName,
+				},
+				Spec: harvesterv1.ResourceQuotaSpec{
+					SnapshotLimit: harvesterv1.SnapshotLimit{
+						VMTotalSnapshotSizeQuota: map[string]int64{
+							name: totalSnapshotSizeQuotaInt64,
+						},
+					},
+				},
+			}
+			_, err = h.resourceQuotaClient.Create(resourceQuota)
+			return err
+		}
+		return err
+	}
+
+	resourceQuotaCopy := resourceQuota.DeepCopy()
+	if resourceQuotaCopy.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota == nil {
+		resourceQuotaCopy.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota = map[string]int64{}
+	}
+	resourceQuotaCopy.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota[name] = totalSnapshotSizeQuotaInt64
+	if resourceQuotaCopy.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota[name] == 0 {
+		delete(resourceQuotaCopy.Spec.SnapshotLimit.VMTotalSnapshotSizeQuota, name)
+	}
+	if !reflect.DeepEqual(resourceQuota, resourceQuotaCopy) {
+		_, err = h.resourceQuotaClient.Update(resourceQuotaCopy)
+		return err
+	}
```

Comment:
should we have a separate action to deleteVMSnapshotResourceQuota? rather than deleting if the value is set to 0?